quantus/helpers/model/tf_model.py

```diff
@@ -49,43 +70,50 @@ def __init__(
             softmax=softmax,
             model_predict_kwargs=model_predict_kwargs,
         )
+        self.cache = LRUCache(100)
 
-    def predict(self, x, **kwargs):
-        """
-        Predict on the given input.
-
-        Parameters
-        ----------
-        x: np.ndarray
-            A given input that the wrapped model predicts on.
-        kwargs: optional
-            Keyword arguments.
-
-        Returns
-        --------
-        np.ndarray
-            predictions of the same dimension and shape as the input, values in the range [0, 1].
-        """
-
+    def _get_predict_kwargs(self, **kwargs: Dict[str, ...]) -> Dict[str, ...]:
         # Use kwargs of predict call if specified, but don't overwrite object attribute
-        model_predict_kwargs = {**self.model_predict_kwargs, **kwargs}
-
-        output_act = self.model.layers[-1].activation
-        target_act = activations.softmax if self.softmax else activations.linear
-
-        if output_act == target_act:
-            return self.model(x, training=False, **model_predict_kwargs).numpy()
-
+        all_kwargs = {**self.model_predict_kwargs, **kwargs}
+        # Filter only ones which are supported by Keras
+        predict_kwargs = {
+            k: all_kwargs[k] for k in all_kwargs if k in self._available_predict_kwargs
+        }
+        return predict_kwargs
+
+    @cachedmethod(operator.attrgetter("cache"))
+    def _build_model_with_linear_top(self) -> Model:
+        # In this case model has a softmax on top, and we want linear.
+        # We have to rebuild the model and replace top with linear activation.
         config = self.model.layers[-1].get_config()
-        config["activation"] = target_act
-
+        config["activation"] = activations.linear
```

Comment:
why specifying linear here?

Reply:
If we reach up this part, it means last layers `config["activation"] == activations.softmax`, and we must replace it by 
no activation (called linear in Keras)